crates/services/sync/src/import.rs

```diff
@@ -381,21 +409,62 @@ async fn get_headers_batch(
         .await
         .trace_err("Failed to get headers");
     let sorted_headers = match res {
-        Ok(None) =>
-                vec![Err(anyhow::anyhow!("Headers provider was unable to fulfill request for unspecified reason. Possibly because requested batch size was too large"))],
-        Ok(Some(headers))  =>        headers
+        Ok(sourced_headers) => {
+            let SourcePeer {
+                peer_id,
+                data: maybe_headers,
+            } = sourced_headers;
+            let cloned_peer_id = peer_id.clone();
+            let headers = match maybe_headers {
+                None => {
+                    tracing::error!(
+                        "No headers received from peer {:?} for range {} to {}",
+                        peer_id,
+                        start,
+                        end
+                    );
+                    vec![Err(anyhow::anyhow!("Headers provider was unable to fulfill request for unspecified reason. Possibly because requested batch size was too large"))]
+                }
+                Some(headers) => headers
                     .into_iter()
                     .map(move |header| {
                         let header = range.next().and_then(|height| {
-                            if *(header.data.entity.height()) == height.into() {
-                                Some(header)
+                            if *(header.entity.height()) == height.into() {
+                                let sourced_header = SourcePeer {
+                                    peer_id: cloned_peer_id.clone(),
+                                    data: header,
+                                };
+                                Some(sourced_header)
                             } else {
                                 None
                             }
                         });
                         Ok(header)
                     })
                     .collect(),
+            };
+            if let Some(expected_len) = end.checked_sub(start) {
+                if headers.len() != expected_len as usize
+                    || headers.iter().any(|h| h.is_err())
+                {
+                    let _ = p2p
+                        .report_peer(
+                            peer_id.clone(),
+                            PeerReportReason::MissingBlockHeaders,
+                        )
+                        .await
+                        .map_err(|e| {
+                            tracing::error!(
+                                "Failed to report bad block header from peer {:?}: {:?}",
+                                peer_id,
+                                e
+                            )
+                        });
```

Comment:
If the report fails, it looks like the event is permanently lost. Under what circumstances could this fail?

Reply:
It depends on the implementation of `PeerToPeerPort` we are using for `p2p`. Potentially it could perform 

WRT our specific impl, it looks like there is some fallible code in converting the `peer_id` as well in sending the peer report request.